scipy/special/xsf/cephes/trig.h

```diff
@@ -19,40 +20,66 @@ namespace cephes {
     /* Compute sin(pi * x). */
     template <typename T>
     XSF_HOST_DEVICE T sinpi(T x) {
-        T s = 1.0;
-
-        if (x < 0.0) {
-            x = -x;
-            s = -1.0;
-        }
+        T s = std::copysign(1.0, x);
+        x = std::abs(x);
 
         T r = std::fmod(x, 2.0);
-        if (r < 0.5) {
-            return s * std::sin(M_PI * r);
-        } else if (r > 1.5) {
-            return s * std::sin(M_PI * (r - 2.0));
-        } else {
-            return -s * std::sin(M_PI * (r - 1.0));
-        }
+        s = r <= 1.0 ? s : -s;
+        r = r - std::floor(r);
+        r = r <= 0.5 ? r : (1.0 - r);
+
+        T y = std::copysign(std::sin(M_PI * r), s);
+ 
+        return y;
     }
 
     /* Compute cos(pi * x) */
     template <typename T>
     XSF_HOST_DEVICE T cospi(T x) {
-        if (x < 0.0) {
-            x = -x;
-        }
+        x = std::abs(x);
+
+        T r = std::fmod(x, 2.0);
+        r = r <= 1.0 ? (0.5 - r) : (r - 1.5);
+
+        T y = std::sin(M_PI * r);
+
+        return y;
+    }
+
+    /* Compute tan(pi * x) */
+    template <typename T>
+    XSF_HOST_DEVICE T tanpi(T x) {
+        T s = std::copysign(1.0, x);
+        x = std::abs(x);
 
         T r = std::fmod(x, 2.0);
-        if (r == 0.5) {
-            // We don't want to return -0.0
-            return 0.0;
-        }
-        if (r < 1.0) {
-            return -std::sin(M_PI * (r - 0.5));
-        } else {
-            return std::sin(M_PI * (r - 1.5));
-        }
+        s = r <= 1.0 ? s : -s;
+
+        T rs = r - std::floor(r);
+        rs = rs <= 0.5 ? rs : (1.0 - rs);
+        T rc = r <= 1.0 ? (0.5 - r) : (r - 1.5);
+
+        T y = std::copysign(std::sin(M_PI * rs), s) / std::sin(M_PI * rc);
+
+        return y;
```

Comment:
Would it be better to call `std::tan` instead? That might be more accurate than computing the ratio of two sines,  and probably saves half the computation.

Separately, would it be useful to make a special case for `x = 0.25`? I don't know, just asking.

Reply:
> Would it be better to call `std::tan` instead? That might be more accurate than computing the ratio of two sines, and probably saves half the computation.


> Separately, would it be useful to make a special case for `x = 0.25`? (This is _not_ required by IEEE, and Julia's `tanpi(0.25) < 1`.) It might look nice but it adds a branch and slows down the code for 99.99% cases. I don't know, just asking.

If it's not required by the spec, and other implementations don't special case this, I think it's fine not to bother.